keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+-->
+
+#### Rejected alternatives
+
+##### Portable regex behavior
+
+The safest route for specifying the regex would be to state a limited
+subset that can be used in a portable way. Any expressions outside of
+the subset will have implementation specific behavior.
+
+Regular expression subset (derived from [re2][re2-syntax] syntax page)
+
+| Expression | description             |
+|------------|-------------------------|
+| `.`        | any character           |
+| `[xyz]`    | character class         |
+| `[^xyz]`   | negated character class |
+| `x*`       | 0 or more x's           |
+| `x+`       | 1 or more x's           |
+| `xy`       | x followed by y         |
+| `x|y`      | x or y (prefer x)       |
+| `(abc)`    | grouping                |
+
+Maintaining a regular expression subset is likely not worth the complexity and
+is likely impossible across the [many implementations][regex-survey].
+
+### API field naming
+
+These are straightforwarding one-to-one renames to for consistency.
+
+| v1beta1 field  | ga field              | rationale                   |
+|----------------|-----------------------|-----------------------------|
+| `spec.backend` | `spec.defaultBackend` | Explicitly mentions default |
+
+### Hostname wildcards
+
+Most platforms support wildcarding for hostnames, e.g. syntax such as
+`*.foo.com` matches names `app1.foo.com`, `app2.foo.com`. The current spec
+states that `spec.rules.host` must be a FQDN of a network host. 
+
+#### Proposal
+
+This proposal would be a limited support for adding a single wildcard `*` as the
+first host label.
+
+The `IngressRule.Host` comment would be changed to:
+
+> `Host` can be a "precise host" which is an fully-qualified domain name without
+> the terminating dot of a network host (e.g. "foo.bar.com") or a "wildcard
+> host" domain name prefixed with a single wildcard label ("*.foo.com").
+> Requests will be matched against the `Host` field in the following way:
+> 
+> If `Host` is precise, the request matches the rule if the http host header
+> is equal to `Host`.
+> 
+> If `Host` is a wildcard, then request matches the rule if the http host header
+> is to equal to the suffix (removing the first label) of the wildcard rule.
+> E.g. wildcard "*.foo.com" matches "bar.foo.com" because they share an equal
+> suffix "foo.com".
+
+### Healthchecks
+
+The current spec does not have any provisions to customize appropriate
+healtchecks for referenced backends. Many users already have a healthcheck
+URLthat is lightweight and different from the HTTP root (e.g. `/`).
+
+One option that has been explored is to infer the healthcheck URL from the
+Readiness probes on the Pods of the Service. This method has proven to be quite
+unstable: Every Pod in a Service can have a different Readiness probe definition
+and it's not clear which one should be used. Furthermore, the behavior is quite
+implicit and creates action-at-a-distance relationship between the Ingress and
+Pod resources.
```

Comment:
This is not duplicating the already existing probes in k8s? Why a new health check and not encourage the definition of readiness and liveness probes in the deployments?

Reply:
Is this healthcheck endpoint going to be about a backend service or a specific endpoint of backend service?

Reply:
While I totally agree that pod-probes sort of cover this already, two things are different:

1) end-to-end checking for LBs like this can detect failures that pod probes can not.  E.g a pod could still be probed successfully (by kubelet) while the node itself suffers some network problems.

2) cloud-ish APIs often require a healthcheck be specified, and we have had many many many reports from users that the heuristic we use today (go look at the pods) does not work well.  For example, a Service can point to Pods from multiple Deployments, and those pods might have different health-checks.  This allows people to specify it explicitly.

Reply:
updated text